### PR TITLE
Get terraform files from remote repo

### LINKED ISSUE #4: Source code retrieval from Git
Using the same structures as found in Argo CD, retrieve source code from a Git repository

Current process for source code retrieval

* [GRPC request](https://github.com/argoproj/argo-cd/blob/edc85a3aa8d9488fc96688c88a17d07b825d74f0/reposerver/repository/repository.proto#L159) from application controller -> repo service
GenerateManifests logic in repo server:
1) Receive request from GRPC
2) Clone whatever the GRPC asked us to (repo url, path within repo, branch, credentials, etc)
3) Run kustomize/helm/etc however the GRPC asked us to (kustomize/helm parametes, config options, etc)
4) return the resulting YAML as a string to caller via GRPC response (edited)

[Implementation Code](https://github.com/argoproj/argo-cd/blob/edc85a3aa8d9488fc96688c88a17d07b825d74f0/reposerver/repository/repository.go)

### PR DESCRIPTION
**Pull request by josephsawaya**
This commit implements getting the terraform files
from the remote repository using the argocd-repo-server.

This commit implements an argocd plugin which is necessary
fetching the remote terraform files. This plugin calls a script
which runs in the path specified by the Terraform CR. Similar to 
this [ArgoCD vault plugin](https://github.com/argoproj-labs/argocd-vault-plugin).

This commit creates a RepoClient to call the GenerateManifests
function in the repo server to make use of this plugin and get
the files. Finally it prints the contents of the manifests
returned by the the repo server.

Fixes: #4 

Depends on: #14

### CODE REVIEW

File: api/v1alpha1/terraform_types.go [hunk 1]
```diff
@@ -28,16 +29,45 @@ type TerraformSpec struct {
...
+// path of file containing terraform state
+type TerraformState string
+
+/* This could be:
+Initializing
+Planning
+Applying
+Destroying
+Failed
+*/
```
Comment (jeffvance) [inline api/v1alpha1/terraform_types.go:68]: Should this state constants be defined?

File: controllers/terraform_controller.go [hunk 2]
```diff
@@ -17,13 +17,18 @@ limitations under the License.
 package controllers
 
 import (
-	"context"
+	context "context"
```
Comment (jeffvance) [inline controllers/terraform_controller.go]: why an alias here?

File: Makefile [hunk 3]
```diff
@@ -231,3 +259,19 @@ catalog-build: opm ## Build a catalog image.
...
+set -e ;\
+echo "Downloading $(2)" ;\
+curl -sSLo "$(1)" "$(2)" ;\
+chmod a+x "$(1)" ;\
+}
+endef
+
+.PHONY: kuttl
+KUTTL := $(PROJECT_DIR)/bin/kuttl
+KUTTL_URL := https://github.com/kudobuilder/kuttl/releases/download/v$(KUTTL_VERSION)/kubectl-kuttl_$(KUTTL_VERSION)_linux_x86_64
```
Comment (sabre1041) [inline Makefile]: Binary to retrieve should take account of the platform/operating system that it is being installed on

File: config/manager/manager.yaml [hunk 4]
```diff
@@ -3,13 +3,13 @@ kind: Namespace
 metadata:
   labels:
     control-plane: controller-manager
-  name: system
+  name: argocd
```
Comment (sabre1041) [inline config/manager/manager.yaml]: Lets keep these as the default `system` namespace. Values will be dynamically updated at runtime

File: config/manager/manager.yaml [hunk 5]
```diff
@@ -32,6 +32,7 @@ spec:
         args:
         - --leader-elect
         image: controller:latest
+        imagePullPolicy: Always
```
Comment (sabre1041) [inline config/manager/manager.yaml]: This is good for development, but not needed in production as it should be referencing the digest

File: config/samples/_v1alpha1_terraform.yaml [hunk 6]
```diff
@@ -2,5 +2,10 @@ apiVersion: argoproj.io/v1alpha1
 kind: Terraform
 metadata:
   name: terraform-sample
+  namespace: argocd
 spec:
-  # TODO(user): Add fields here
+  source:
+    repoURL: https://github.com/josephsawaya/terraform-test.git
```
Comment (sabre1041) [inline config/samples/_v1alpha1_terraform.yaml]: Do we want to incorporate the sample into this repository?

File: terraform-generate/kustomize-install/argocd-repo-server.yaml [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+      - name: custom-tools
+        emptyDir: {}
+      initContainers:
+      - name: download-tools
+        image: alpine:3.8
+        command: [sh, -c]
+        args:
+          - >-
+            wget -O argocd-terraform-generator
+            https://raw.githubusercontent.com/josephsawaya/argocd-terraform-controller/dev/terraform-generate/generate.sh &&
```
Comment (sabre1041) [inline terraform-generate/kustomize-install/argocd-repo-server.yaml]: This needs to be updated for this repository

**Review by sabre1041 — CHANGES_REQUESTED** [review]
Initial feedback

File: Makefile [hunk 8]
```diff
@@ -231,3 +259,19 @@ catalog-build: opm ## Build a catalog image.
...
+define download-tool
+@[ -f $(1) ] || { \
+set -e ;\
+echo "Downloading $(2)" ;\
+curl -sSLo "$(1)" "$(2)" ;\
+chmod a+x "$(1)" ;\
+}
+endef
+
+.PHONY: kuttl
```
Comment (sabre1041) [inline Makefile]: Download fails if `bin` folder not present first

File: Makefile (hunk 3 above)
Comment (sabre1041) [reply Makefile]: Heres an example:

```
# Platform
ifeq ($(OS),Windows_NT)
  PLATFORM := windows
else
  UNAME_S = $(shell uname -s)
  ifeq ($(UNAME_S),Linux)
    PLATFORM := linux
  endif
  ifeq ($(UNAME_S),Darwin)
    PLATFORM := darwin
  endif
endif
```

`$(PLATFORM)` can then be added to the `KUTTL_URL`

File: Makefile [hunk 9]
```diff
@@ -106,6 +108,11 @@ vet: ## Run go vet against code.
 test: manifests generate fmt vet envtest ## Run tests.
 	KUBEBUILDER_ASSETS="$(shell $(ENVTEST) use $(ENVTEST_K8S_VERSION) -p path)" go test ./... -coverprofile cover.out
 
+.PHONY: test-e2e
+test-e2e: kuttl ## Run e2e tests. Requires cluster w/ Scribe already installed
```
Comment (sabre1041) [inline Makefile]: add `test-kuttl/kubeconfig` to `.gitignore`

**Review by sabre1041 — CHANGES_REQUESTED** [review]
I tried to run the test framework without success. Are there steps that need to be followed? Add to `README.md`?

**Comment by josephsawaya** [comment]
> I tried to run the test framework without success. Are there steps that need to be followed? Add to `README.md`?

So the test framework changes are a part of #13. I realize now that having many PRs open that depend on each other is probably confusing, so I'm going to close my previous PR, and try to remove the commits in this PR that are a part of #13.

**Review by sabre1041 — APPROVED** [review]
LGTM

**Pull request merged**